### PR TITLE
[FLINK-16812][jdbc] support array types in PostgresRowConverter

### PR DESCRIPTION
**Pull request by bowenli86**
## What is the purpose of the change

commit 1: fix bug that StringUtils.arrayToString() should consider Object[] lastly, otherwise it will take byte[][] as Object[] and produce wrong string
commit 2: support array types in PostgresRowConverter

## Verifying this change

with UTs added

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

docs will added later

### CODE REVIEW

**Comment by bowenli86** [comment]
@wuchong @huafengw pls take a look

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java [hunk 1]
```diff
@@ -28,4 +34,39 @@
...
+
+					Object[] out = new Object[in.length];
+					for (int i = 0; i < in.length; i++) {
+						out[i] = ((PGobject) in[i]).getValue().getBytes();
+					}
+
+					return out;
+				};
+			} else {
+				return v -> ((PgArray) v).getArray();
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java:60]: Should we add the default conversion for ARRAY in `AbstractJDBCRowConverter`? Currently, we will directly put `java.sql.Array` into the Row which is not correct.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java [hunk 2]
```diff
@@ -28,4 +34,39 @@
...
+			if (elemType == LogicalTypeRoot.VARBINARY) {
+
+				return v -> {
+					PgArray pgArray = (PgArray) v;
+					Object[] in = (Object[]) pgArray.getArray();
+
+					Object[] out = new Object[in.length];
+					for (int i = 0; i < in.length; i++) {
+						out[i] = ((PGobject) in[i]).getValue().getBytes();
+					}
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java:55]: Could you add a comment why PG should special handle `ARRAY<VARBINARY>`?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java [hunk 3]
```diff
@@ -28,4 +34,39 @@
...
+
+	@Override
+	public JDBCFieldConverter createConverter(LogicalType type) {
+		LogicalTypeRoot root = type.getTypeRoot();
+
+		if (root == LogicalTypeRoot.ARRAY) {
+			ArrayType arrayType = (ArrayType) type;
+			LogicalTypeRoot elemType = arrayType.getElementType().getTypeRoot();
+
+			if (elemType == LogicalTypeRoot.VARBINARY) {
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java:46]: I would suggest to use `LogicalTypeChecks#hasFamily(elemType, LogicalTypeFamily.BINARY_STRING)` to also support `BINARY`.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java (hunk 1 above)
Comment (bowenli86) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java:60]: >Should we add the default conversion for ARRAY in AbstractJDBCRowConverter? 
>Currently, we will directly put java.sql.Array into the Row which is not correct.

There's no default array conversion yet. In pg, it's always need to case to `PgArray`. 

What it puts in a Row is primitive array (obtained from PgArray.getArray()), not java.sql.Array

Comment (wuchong) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java:60]: I mean the currently default implementation of `AbstractJDBCRowConverter` will use `v -> v` for the array conversion, which puts `java.sql.Array` in a Row.

Comment (bowenli86) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/source/row/converter/PostgresRowConverter.java:60]: gotcha. 

AbstractJDBCRowConverter is migrated from JDBCInputFormat, It's always been like that in JDBCInputFormat. We can explicitly disable array in AbstractJDBCRowConverter in a separate jira if you want.

**Comment by bowenli86** [comment]
@wuchong addressed comments. pls take another look

**Pull request closed without merging**